# **Segmenting and Clustering Neighbourhoods in Toronto**

##Importing data

In [3]:
import pandas as pd
import numpy as np
import requests

Get the HTML page of Wiki, and using read_html we convert the html data into list of Data frame objects.


In [4]:


wiki = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
wiki_page = requests.get(wiki)

wiki_data = pd.read_html(wiki_page.content, header = 0)[0] #We need the first table alone, so dropping the other tables
wiki_data.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

Remove cells which have borrough not assigned

In [5]:
df = wiki_data[wiki_data["Borough"] != "Not assigned"]
df

Postal Code  ...                                      Neighbourhood
2           M3A  ...                                          Parkwoods
3           M4A  ...                                   Victoria Village
4           M5A  ...                          Regent Park, Harbourfront
5           M6A  ...                   Lawrence Manor, Lawrence Heights
6           M7A  ...        Queen's Park, Ontario Provincial Government
..          ...  ...                                                ...
160         M8X  ...      The Kingsway, Montgomery Road, Old Mill North
165         M4Y  ...                               Church and Wellesley
168         M7Y  ...  Business reply mail Processing Centre, South C...
169         M8Y  ...  Old Mill South, King's Mill Park, Sunnylea, Hu...
178         M8Z  ...  Mimico NW, The Queensway West, South of Bloor,...

[103 rows x 3 columns]

Grouping the records based on Postal Code

In [6]:
df = df.groupby(['Postal Code']).head()
df

Postal Code  ...                                      Neighbourhood
2           M3A  ...                                          Parkwoods
3           M4A  ...                                   Victoria Village
4           M5A  ...                          Regent Park, Harbourfront
5           M6A  ...                   Lawrence Manor, Lawrence Heights
6           M7A  ...        Queen's Park, Ontario Provincial Government
..          ...  ...                                                ...
160         M8X  ...      The Kingsway, Montgomery Road, Old Mill North
165         M4Y  ...                               Church and Wellesley
168         M7Y  ...  Business reply mail Processing Centre, South C...
169         M8Y  ...  Old Mill South, King's Mill Park, Sunnylea, Hu...
178         M8Z  ...  Mimico NW, The Queensway West, South of Bloor,...

[103 rows x 3 columns]

In [7]:
df = df.reset_index()
df

index  ...                                      Neighbourhood
0        2  ...                                          Parkwoods
1        3  ...                                   Victoria Village
2        4  ...                          Regent Park, Harbourfront
3        5  ...                   Lawrence Manor, Lawrence Heights
4        6  ...        Queen's Park, Ontario Provincial Government
..     ...  ...                                                ...
98     160  ...      The Kingsway, Montgomery Road, Old Mill North
99     165  ...                               Church and Wellesley
100    168  ...  Business reply mail Processing Centre, South C...
101    169  ...  Old Mill South, King's Mill Park, Sunnylea, Hu...
102    178  ...  Mimico NW, The Queensway West, South of Bloor,...

[103 rows x 4 columns]

In [8]:
df.drop(['index'], axis = 'columns', inplace = True)
df

Postal Code  ...                                      Neighbourhood
0           M3A  ...                                          Parkwoods
1           M4A  ...                                   Victoria Village
2           M5A  ...                          Regent Park, Harbourfront
3           M6A  ...                   Lawrence Manor, Lawrence Heights
4           M7A  ...        Queen's Park, Ontario Provincial Government
..          ...  ...                                                ...
98          M8X  ...      The Kingsway, Montgomery Road, Old Mill North
99          M4Y  ...                               Church and Wellesley
100         M7Y  ...  Business reply mail Processing Centre, South C...
101         M8Y  ...  Old Mill South, King's Mill Park, Sunnylea, Hu...
102         M8Z  ...  Mimico NW, The Queensway West, South of Bloor,...

[103 rows x 3 columns]

In [9]:
df.head(12)

Postal Code  ...                                      Neighbourhood
0          M3A  ...                                          Parkwoods
1          M4A  ...                                   Victoria Village
2          M5A  ...                          Regent Park, Harbourfront
3          M6A  ...                   Lawrence Manor, Lawrence Heights
4          M7A  ...        Queen's Park, Ontario Provincial Government
5          M9A  ...            Islington Avenue, Humber Valley Village
6          M1B  ...                                     Malvern, Rouge
7          M3B  ...                                          Don Mills
8          M4B  ...                    Parkview Hill, Woodbine Gardens
9          M5B  ...                           Garden District, Ryerson
10         M6B  ...                                          Glencairn
11         M9B  ...  West Deane Park, Princess Gardens, Martin Grov...

[12 rows x 3 columns]

print the number of rows

In [10]:
rows = len(df['Postal Code'].unique())
print("Total number of rows is", rows)

Total number of rows is 103


#**Part 2**

In [11]:
pip install geocoder

     |████████████████████████████████| 102kB 2.3MB/s 


In [12]:
import geocoder

geocoder wasn't working that's why using csv file given in the course

In [13]:
url = 'http://cocl.us/Geospatial_data'

In [14]:
df_geo = pd.read_csv(url)
df_geo.head(12)

Postal Code   Latitude  Longitude
0          M1B  43.806686 -79.194353
1          M1C  43.784535 -79.160497
2          M1E  43.763573 -79.188711
3          M1G  43.770992 -79.216917
4          M1H  43.773136 -79.239476
5          M1J  43.744734 -79.239476
6          M1K  43.727929 -79.262029
7          M1L  43.711112 -79.284577
8          M1M  43.716316 -79.239476
9          M1N  43.692657 -79.264848
10         M1P  43.757410 -79.273304
11         M1R  43.750072 -79.295849

In [15]:
print("The shape of our wiki data is: ", df.shape)
print("the shape of our csv data is: ", df_geo.shape)

The shape of our wiki data is:  (103, 3)
the shape of our csv data is:  (103, 3)


Since the dimensions are the same, we can try to join on the postal codes to get the required data.

Checking the column types of both the dataframes, especially Postal Code column since we are trying to join on it

In [16]:
df.dtypes

Postal Code      object
Borough          object
Neighbourhood    object
dtype: object

In [18]:
df_geo.dtypes

Postal Code     object
Latitude       float64
Longitude      float64
dtype: object

In [21]:
combined_data = df.join(df_geo.set_index('Postal Code'), on='Postal Code', how='inner')
combined_data.head(12)

Postal Code           Borough  ...   Latitude  Longitude
0          M3A        North York  ...  43.753259 -79.329656
1          M4A        North York  ...  43.725882 -79.315572
2          M5A  Downtown Toronto  ...  43.654260 -79.360636
3          M6A        North York  ...  43.718518 -79.464763
4          M7A  Downtown Toronto  ...  43.662301 -79.389494
5          M9A         Etobicoke  ...  43.667856 -79.532242
6          M1B       Scarborough  ...  43.806686 -79.194353
7          M3B        North York  ...  43.745906 -79.352188
8          M4B         East York  ...  43.706397 -79.309937
9          M5B  Downtown Toronto  ...  43.657162 -79.378937
10         M6B        North York  ...  43.709577 -79.445073
11         M9B         Etobicoke  ...  43.650943 -79.554724

[12 rows x 5 columns]

In [20]:
combined_data.shape

(103, 5)

We get 103 rows as expected when we do a inner join, so we have good data.